idseq_dag/steps/generate_phylo_tree.py

```diff
@@ -82,33 +90,60 @@ def run(self):
+                    f"kSNP3 -in inputs.txt -outdir ksnp3_outputs -k 13")
+        if os.path.isfile(annotated_genome_input):
+            ksnp_cmd += f" -annotate {os.path.basename(annotated_genome_input)}"
+            # Note: produces SNP annotation file in a human-readable format that's very space inefficient (~100 MB).
+            # May want to do some postprocessing once we know better what exactly we need for the web app.
+        command.execute(ksnp_cmd)
         command.execute(f"mv {self.output_dir_local}/ksnp3_outputs/tree.parsimony.tre {output_files[0]}")
+        annotation_output = f"{self.output_dir_local}/ksnp3_outputs/SNPs_all_annotated"
+        if os.path.isfile(annotation_output):
+            try:
+                command.execute(f"mv  {self.output_dir_local}/ksnp3_outputs/SNPs_all_annotated {output_files[1]}")
+            except:
+                # HACK because web still makes the old DAG. Remove after new DAG is deployed.
+                command.execute(f"aws s3 cp {self.output_dir_local}/ksnp3_outputs/SNPs_all_annotated {os.path.join(self.output_dir_s3, 'SNP_annotations.txt')}")
+        else:
+            log.write(f"Warning: no SNP annotation file was produced!")
 
     def count_reads(self):
         pass
 
-    def get_genbank_genomes(self, taxid, destination_dir, n=10):
+    def get_genbank_genomes(self, reference_taxids, destination_dir, superkingdom_name, n=10):
         '''
-        Retrieve up to n GenBank reference genomes under taxid.
-        Assumes taxid is species-level.
+        Retrieve up to n GenBank reference genomes under the reference_taxids.
+        Assumes reference_taxids are species-level or below.
+        Also assumes they are all in the same superkingdom, which is the only thing we need in our application.
         Saves the references under file names compatible with MakeKSNP3infile.
         '''
-        if n == 0:
+        if n == 0 or not reference_taxids:
             return []
-        categories = ["bacteria", "viral", "fungi", "protozoa"]
-        # additional options in genbank that probably don't need right now:
+        n_per_taxid = max(n // len(reference_taxids), 1)
+        genbank_categories_by_superkingdom = {
+            "Viruses": ["viral"],
+            "Bacteria": ["bacteria"],
+            "Eukaryota": ["fungi", "protozoa"],
+            None: ["bacteria", "viral", "fungi", "protozoa"]
+        }
+        # additional options in genbank that we probably don't need right now:
         # ["archaea", "plant", 
         # "vertebrate_mammalian", "vertebrate_other", "invertebrate",
         # "other", "metagenomes"]
+        categories = genbank_categories_by_superkingdom[superkingdom_name]
         for cat in categories:
             genome_list_path = f"ftp://ftp.ncbi.nih.gov/genomes/genbank/{cat}/assembly_summary.txt"
             genome_list_local = f"{destination_dir}/{os.path.basename(genome_list_path)}"
```

Comment:
so we are downloading the genome list every time you do this, per our discussion, maybe it would be more efficient and robust to ask jonathan to setup a synch to get the list and find the accession ids and query our nt/nr database?

Reply:
ok. I asked Jonathan to set up a sync. I will also put the genome list in S3 already so the only thing we will have to query ncbi for is the actual genome. We could use our own nt_loc.db, but there is a missing link: assembly_summary.txt contains GenBank accessions like GCA_000010525.1 and we need to figure out how to convert it to NT accessions like AP009384.1. E.g. see https://www.ncbi.nlm.nih.gov/assembly/GCF_000010525.1/. I'll make a task for it.

Reply:
https://jira.czi.team/browse/IDSEQ-334